include/crocoddyl/core/codegen/action-base.hpp

```diff
@@ -0,0 +1,299 @@
+
+  typedef typename PINOCCHIO_EIGEN_PLAIN_ROW_MAJOR_TYPE(ADMatrixXs) RowADMatrixXs;
+
+  typedef CppAD::ADFun<CGScalar> ADFun;
+
+  ActionModelCodeGenTpl(boost::shared_ptr<ADBase> admodel, boost::shared_ptr<Base> model,
+                        const std::string& library_name, const std::string& function_name_calc = "calc",
+                        const std::string& function_name_calcDiff = "calcDiff")
+      : Base(model->get_state(), model->get_nu()),
+        model(model),
+        ad_model(admodel),
+        ad_X(ad_model->get_state()->get_nx() + ad_model->get_nu()),
+        ad_X2(ad_model->get_state()->get_nx() + ad_model->get_nu()),
+        ad_calcout(ad_model->get_state()->get_nx() + 1),
+        ad_data(ad_model->createData()),
+        function_name_calc(function_name_calc),
+        function_name_calcDiff(function_name_calcDiff),
+        library_name(library_name) {
+    const std::size_t& ndx = ad_model->get_state()->get_ndx();
+    const std::size_t& nu = ad_model->get_nu();
+    ad_calcDiffout.resize(2 * ndx * ndx + 2 * ndx * nu + nu * nu + ndx + nu);
+    initLib();
+    loadLib();
+  }
+  void recordCalc() {
+    CppAD::Independent(ad_X);
+    ad_model->calc(ad_data, ad_X.head(ad_model->get_state()->get_nx()), ad_X.tail(ad_model->get_nu()));
+    collect_calcout();
+    // ad_calcout.template head<1>()[0] = ad_data->cost;
+    // ad_calcout.tail(ad_model->get_state()->get_nx()) = ad_data->xnext;
+    ad_calc.Dependent(ad_X, ad_calcout);
+    ad_calc.optimize("no_compare_op");
+  }
+
+  void collect_calcout() {
+    ad_calcout[0] = ad_data->cost;
+    ad_calcout.tail(ad_model->get_state()->get_nx()) = ad_data->xnext;
+  }
+
+  void collect_calcDiffout() {
+    ADVectorXs& ad_Y = ad_calcDiffout;
+
+    const std::size_t& ndx = ad_model->get_state()->get_ndx();
+    const std::size_t& nu = ad_model->get_nu();
+    Eigen::DenseIndex it_Y = 0;
+    Eigen::Map<ADMatrixXs>(ad_Y.data() + it_Y, ndx, ndx) = ad_data->Fx;
+    it_Y += ndx * ndx;
+    Eigen::Map<ADMatrixXs>(ad_Y.data() + it_Y, ndx, nu) = ad_data->Fu;
+    it_Y += ndx * nu;
+    Eigen::Map<ADVectorXs>(ad_Y.data() + it_Y, ndx) = ad_data->Lx;
+    it_Y += ndx;
+    Eigen::Map<ADVectorXs>(ad_Y.data() + it_Y, nu) = ad_data->Lu;
+    it_Y += nu;
+    Eigen::Map<ADMatrixXs>(ad_Y.data() + it_Y, ndx, ndx) = ad_data->Lxx;
+    it_Y += ndx * ndx;
+    Eigen::Map<ADMatrixXs>(ad_Y.data() + it_Y, ndx, nu) = ad_data->Lxu;
+    it_Y += ndx * nu;
+    Eigen::Map<ADMatrixXs>(ad_Y.data() + it_Y, nu, nu) = ad_data->Luu;
+  }
+
+  void recordCalcDiff() {
+    CppAD::Independent(ad_X2);
+    // ActionDataCodeGenTpl<ADScalar>* ad_d =
+    // static_cast<ActionDataCodeGenTpl<ADScalar>*>(ad_data.get());
+    ad_model->calc(ad_data, ad_X2.head(ad_model->get_state()->get_nx()), ad_X2.tail(ad_model->get_nu()));
+    ad_model->calcDiff(ad_data, ad_X2.head(ad_model->get_state()->get_nx()), ad_X2.tail(ad_model->get_nu()));
+
+    collect_calcDiffout();
```

Comment:
We could. But we don't use it, since the calc output is being already generated by the calc function, and we don't use calcdiff-with-calc anymore. So I don't think we need to have additional outputs.

However, the full API would be much more clearer once we start to put this in use.